### ISSUE TITLE
Allow "Compatible" Scala Versions in Cross-Build Mechanism

### ISSUE THREAD

**Issue by djspiewak** [sbt/sbt#295]
Cross-builds are a lovely hack that do a very good job of resolving many of the common cases of Scala version incompatibilities across dependencies.  Unfortunately, they also couple libraries together in a very unpleasant way.  This is particularly apparent when a new version of Scala lands and the entire community needs to perform a topological walk of their libraries, pushing new cross-builds.  Projects that are dependencies for numerous other projects (think: Specs and ScalaCheck) become blockers for the entire community, since no one can push an updated cross-build for their own library before their dependencies.  This has given rise to an extremely common pattern of dependency specification, where a project will use the cross-build artifact id for a particular library, but only for certain versions of Scala, using a custom id for other version (generally the latest one).

In theory, minor releases of Scala are now backwards compatible.  Thus, cross-builds that were built against 2.9.0 _should_ be compatible with dependant libraries being built against 2.9.1 (and vice versa).  We certainly can't guarantee this in general, but it's a safe rule of thumb.  In any case, it should be possible to take advantage of this fact without having to manually override cross-build artifact identifiers.

I am proposing a new property, `compatibleScalaVersions` of type `Seq[String]`.  This property would default to simply `Seq(scalaVersion)` for backwards compatibility with prior SBT semantics.  The functionality of this property would be visible in the cross-build semantics.  Specifically, if the value of `scalaVersion` is `"2.9.1"` and the project is dependent on some artifact `foo` that has a cross-build for `2.9.0` but not `2.9.1`, a build could specify `compatibleScalaVersion += "2.9.0"` and the cross-build of `foo` for `2.9.0` would be selected in lieu of a more specific cross-build for `2.9.1`.  When a cross-build for `scalaVersion` is available, it will always be selected.  Otherwise, SBT should attempt to resolve cross-builds for each value in `compatibleScalaVersions` in order.  This would give library and project authors a controlled and declarative way of specifying compatible Scala versions, effectively relaxing the (currently lexical) constraints of cross-build resolution.

There are some issues here, such as the fact that `compatibleScalaVerions` includes `scalaVersion`, which makes the fallback semantics a little strange.  It might be better to just implicitly assume that `scalaVersion` will _always_ be tried first, defaulting `compatibleScalaVersions` to `Seq()`.

**Comment by jrudolph** [comment]
Here's some previous discussion about this topic: 

https://groups.google.com/d/topic/simple-build-tool/z4ubUB_QbF8/discussion

**Comment by djspiewak** [comment]
The partial function idea on that thread is interesting, but non-declarative and bulky.  I think the `compatibleScalaVersions` is a nicer abstraction.  It's effectively the "list of compatible versions in SBT" idea, except controlled by build authors rather than baked into the tool.

**Comment by rossabaker** [comment]
I have two concerns with this approach:

1) What goes into the POM?  Most transitive dependencies are acquired via POM, even those built with SBT.  Suppose `foo_2.9.1` releases against `quux_2.9.0`.  You depend on `foo_2.9.1` and the freshly published `quux_2.9.1`.  Nothing in foo's POM mentions compatible versions, so now you have two quuxes in your classpath.  Dependency managers could be trained that `quux_2.9.1` supersedes `quux_2.9.0`.  But this is tricky, because crossbuilds introduce a second axis of versioning.  If a project inherits both `quux_2.9.1-1.0` and `quux_2.9.0-1.1`, which one wins?

2) It makes your build non-repeatable.  A tagged release should always depend on a fixed set of dependencies.  Using this mechanism, my release may build against different libraries next week than it does today, without me changing my build.  This is the same as depending on a snapshot.  It's wonderful for staying current during development, but not suitable for a stable release.

**Comment by odersky** [comment]
I think it's high time to put this into the build. In principle we now check that minor versions are binary compatible with each other, so compatibleScalaVersions should be initialized by default to reflect that.

**Comment by djspiewak** [comment]
> 1) What goes into the POM?

Does anyone know if Maven2 version range specifiers also work with artifact ids?  That would certainly be the ideal situation.  If this isn't possible, then we have a bit of a problem because of the two-axis version issue.

> 2) It makes your build non-repeatable.

As long as people aren't abusing cross-builds to change functionality between Scala versions, this is a non-issue.  One of the major "cross-build assumptions" is that libraries are equivalent between versions of Scala, just rebuilt.  So in other words, your build is still repeatable, it will just have a different set of nominal dependencies.  Depending on how we resolve issue (1) in your list, this should be a workable solution.

> In principle we now check that minor versions are binary compatible with each other, so compatibleScalaVersions should be initialized by default to reflect that. 

As long as it's still possible to override the default `compatibleScalaVersions` on a per-build basis, I'm fine with initializing its value to the default compatibility set.  Sane defaults are good.  The one thing is that we need to think about where we get that default compatibility set for each version of Scala.  Hard-coding compatibility sets into SBT itself seems like a bad idea.

**Comment by rossabaker** [comment]
> > 1) What goes into the POM?
> 
> Does anyone know if Maven2 version range specifiers also work with artifact ids? That would certainly be the ideal situation. If this isn't possible, then we have a bit of a problem because of the two-axis version issue.

I believe not.  There's no notion of comparability for artifact ids, so a range would make no sense in that field.  The Scala community using it as a second versioning axis is a hack.  A nifty hack, but a hack.

> > 2) It makes your build non-repeatable.
> 
> As long as people aren't abusing cross-builds to change functionality between Scala versions, this is a non-issue. One of the major "cross-build assumptions" is that libraries are equivalent between version1ns of Scala, just rebuilt. So in other words, your build is still repeatable, it will just have a different set of nominal dependencies. Depending on how we resolve issue (1) in your list, this should be a workable solution.

I agree that should be part of the cross-build contract.  But there's a significant gap between trusting a known, fully tested build vs. trusting all future cross-builds to .honor to this contract.  Verifying the compatibility of the dependencies is my responsibility as a developer, and I can't continue that responsibility after I hand off the release.

**Comment by indrajitr** [comment]
Quite likely we'll have something to deal with this in 0.12.

TL;DR you can overload `%%` to take an optional second argument which can be of the from `String => String` or a bunch of `(String,String)*`.

See 0270ac078ff23adf9b1edc6d5e9ffd52ff9059d1 and the related discussion #267 for details.

FWIW, despite being on 0.11 branch for about a month, it didn't make it to 0.11.2 because the intent was to keep it strictly a _bug fix_ release.

**Comment by rossabaker** [comment]
+1 to @harrah's solution in #267:

> Something that I'd like to see done soon is drop the incremental version for Scala 2.10 and higher.

This problem recedes once the naming convention catches up to the advances in binary compatibility.  Under that new convention, libraries should be built against 2.x.0 to avoid depending on signatures introduced in patch releases.  I already have a scar from using scalate_2.8.2 on 2.8.1.

**Comment by Atry** [comment]
Since scala's source compatibility is better than binary compatibility, we should build another source repository instead of the maven one. With a source repository, people download source, set compiler flag, and compile on their own computer. They can hack the downloaded source files, too.

We don't need port Java projects in Maven, we only need release our Scala project in new source repository.

**Comment by harrah** [comment]
This has been discussed elsewhere, with the resulting solution being the one mentioned in #267 (which I believe @rossabaker originally proposed).  It's already been implemented (but not committed) and will be in 0.12.0.  It will also be applied to sbt plugins.

Please use the mailing list for further discussion.
